### ISSUE TITLE
OpenCL Intel Crash + Windows

### ISSUE THREAD

**Issue by Laurae2** [Microsoft/LightGBM#447]
## Environment info
Operating System: Windows Server 2012 R2
CPU: Dual Intel Xeon unnamed Ivy Bridge (40 virtual cores)
C++/Python/R version: CLI + MinGW 6.3, happens also with 4.9

## Error Message:

Crash

## Reproducible examples

Debugging example

## Steps to reproduce

Run debugging example

Installed: OpenCL Runtime + Intel OpenCL SDK

Tried: system DLL and Intel OpenCL SDK DLL

## CLI Debug Log:

Crash happens between gpu_tree_learner.cpp:572 and gpu_tree_learner.cpp:581 (checked using `breakpoint` and `stepi`), but unknown when it happens exactly (so much instructions it never ends). I commented `pragma omp parallel` at line gpu_tree_learner:569 to allow debugging that point.

I could read manually `invalid program` while using `breakpoint` and `stepi`, it is not depicted below as the log would be gigantic.

Crashes even without `USE_OPENMP` (I removed all omp where appropriate also, it wouldn't compile without)

```
C:\github_repos\LightGBM\examples\binary_classification>gdb --args "../../lightg
bm.exe" config=train.conf data=binary.train valid=binary.test objective=binary d
evice=gpu
GNU gdb (GDB) 7.11.1
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-w64-mingw32".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from ../../lightgbm.exe...done.
(gdb) run
Starting program: C:\github_repos\LightGBM\lightgbm.exe "config=train.conf" "dat
a=binary.train" "valid=binary.test" "objective=binary" "device=gpu"
[New Thread 4048.0xf34]
[New Thread 4048.0xfcc]
[Thread 4048.0xfcc exited with code 0]
[LightGBM] [Info] Finished loading parameters
[New Thread 4048.0x2b8]
[New Thread 4048.0x668]
[New Thread 4048.0x5e4]
[New Thread 4048.0x660]
[New Thread 4048.0x980]
[New Thread 4048.0xfb8]
[New Thread 4048.0x454]
[New Thread 4048.0xa6c]
[New Thread 4048.0xf94]
[New Thread 4048.0xf10]
[New Thread 4048.0x7b0]
[New Thread 4048.0xf8c]
[New Thread 4048.0xd64]
[New Thread 4048.0x97c]
[New Thread 4048.0xf8]
[New Thread 4048.0x614]
[New Thread 4048.0x3b0]
[New Thread 4048.0xf9c]
[New Thread 4048.0x124]
[New Thread 4048.0x610]
[New Thread 4048.0xd54]
[New Thread 4048.0x750]
[New Thread 4048.0x238]
[New Thread 4048.0x76c]
[New Thread 4048.0x96c]
[New Thread 4048.0x900]
[New Thread 4048.0x940]
[New Thread 4048.0x50c]
[New Thread 4048.0xdd4]
[New Thread 4048.0x5f8]
[New Thread 4048.0xe2c]
[New Thread 4048.0x41c]
[New Thread 4048.0x638]
[New Thread 4048.0xaa8]
[New Thread 4048.0x79c]
[New Thread 4048.0xcec]
[New Thread 4048.0xa70]
[New Thread 4048.0x58c]
[New Thread 4048.0xd3c]
[New Thread 4048.0xdac]
[Thread 4048.0xdac exited with code 0]
[LightGBM] [Info] Loading weights...
[New Thread 4048.0x7f4]
[Thread 4048.0x7f4 exited with code 0]
[New Thread 4048.0x4b4]
[Thread 4048.0x4b4 exited with code 0]
[LightGBM] [Info] Loading weights...
[New Thread 4048.0x7e0]
[Thread 4048.0x7e0 exited with code 0]
[LightGBM] [Info] Finished loading data in 1.015635 seconds
[LightGBM] [Info] Number of positive: 3716, number of negative: 3284
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 6143
[LightGBM] [Info] Number of data: 7000, number of used features: 28
[LightGBM] [Info] Using GPU Device: Intel Xeon E3-12xx v2 (Ivy Bridge), Vendor:
Intel(R) Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[New Thread 4048.0x284]
[New Thread 4048.0x91c]
[New Thread 4048.0x7ec]
[New Thread 4048.0x948]
[New Thread 4048.0xa9c]
[New Thread 4048.0x7e8]
Met Exceptions:
Invalid Program
[Thread 4048.0x2b8 exited with code 4294967295]
[Thread 4048.0x660 exited with code 4294967295]
[Thread 4048.0x668 exited with code 4294967295]
[Thread 4048.0x5e4 exited with code 4294967295]
[Thread 4048.0x3b0 exited with code 4294967295]
[Thread 4048.0x91c exited with code 4294967295]
[Thread 4048.0x284 exited with code 4294967295]
[Thread 4048.0xd3c exited with code 4294967295]
[Thread 4048.0x948 exited with code 4294967295]
[Thread 4048.0x7e8 exited with code 4294967295]
[Thread 4048.0x7ec exited with code 4294967295]
[Thread 4048.0x638 exited with code 4294967295]
[Thread 4048.0xe2c exited with code 4294967295]
[Thread 4048.0x96c exited with code 4294967295]
[Thread 4048.0xa9c exited with code 4294967295]
[Thread 4048.0x41c exited with code 4294967295]
[Thread 4048.0x238 exited with code 4294967295]
[Thread 4048.0x750 exited with code 4294967295]
[Thread 4048.0x124 exited with code 4294967295]
[Thread 4048.0xf9c exited with code 4294967295]
[Thread 4048.0x76c exited with code 4294967295]
[Thread 4048.0x58c exited with code 4294967295]
[Thread 4048.0xcec exited with code 4294967295]
[Thread 4048.0xa70 exited with code 4294967295]
[Thread 4048.0x79c exited with code 4294967295]
[Thread 4048.0xd54 exited with code 4294967295]
[Thread 4048.0x900 exited with code 4294967295]
[Thread 4048.0xaa8 exited with code 4294967295]
[Thread 4048.0xf8c exited with code 4294967295]
[Thread 4048.0xf10 exited with code 4294967295]
[Thread 4048.0x980 exited with code 4294967295]
[Thread 4048.0xdd4 exited with code 4294967295]
[Thread 4048.0xf94 exited with code 4294967295]
[Thread 4048.0xfb8 exited with code 4294967295]
[Thread 4048.0x454 exited with code 4294967295]
[Thread 4048.0xd64 exited with code 4294967295]
[Thread 4048.0x50c exited with code 4294967295]
[Thread 4048.0xa6c exited with code 4294967295]
[Thread 4048.0x614 exited with code 4294967295]
[Thread 4048.0x610 exited with code 4294967295]
[Thread 4048.0xf8 exited with code 4294967295]
[Thread 4048.0x940 exited with code 4294967295]
[Thread 4048.0x97c exited with code 4294967295]
[Thread 4048.0x7b0 exited with code 4294967295]
[Thread 4048.0x5f8 exited with code 4294967295]
[Inferior 1 (process 4048) exited with code 037777777777]
```

Without OpenMP:

```
(gdb) run
Starting program: C:\github_repos\LightGBM\lightgbm.exe "config=train.conf" "dat
a=binary.train" "valid=binary.test" "objective=binary" "device=gpu"
[New Thread 264.0x2d8]
[New Thread 264.0xda0]
[Thread 264.0xda0 exited with code 0]
[LightGBM] [Info] Finished loading parameters
[New Thread 264.0xce0]
[Thread 264.0xce0 exited with code 0]
[LightGBM] [Info] Loading weights...
[New Thread 264.0xf08]
[Thread 264.0xf08 exited with code 0]
[New Thread 264.0xa94]
[Thread 264.0xa94 exited with code 0]
[LightGBM] [Info] Loading weights...
[New Thread 264.0xd7c]
[Thread 264.0xd7c exited with code 0]
[LightGBM] [Info] Finished loading data in 1.093739 seconds
[LightGBM] [Info] Number of positive: 3716, number of negative: 3284
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 6143
[LightGBM] [Info] Number of data: 7000, number of used features: 28
[LightGBM] [Info] Using GPU Device: Intel Xeon E3-12xx v2 (Ivy Bridge), Vendor:
Intel(R) Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[New Thread 264.0xd78]
[New Thread 264.0x114]
[New Thread 264.0x790]
[New Thread 264.0x614]
Met Exceptions:
Invalid Program
[Thread 264.0x790 exited with code 4294967295]
[Thread 264.0xd78 exited with code 4294967295]
[Thread 264.0x114 exited with code 4294967295]
[Thread 264.0x614 exited with code 4294967295]
[Inferior 1 (process 264) exited with code 037777777777]
```

ping @huanzhang12

**Comment by huanzhang12** [comment]
The Intel OpenCL runtime complains about the GPU kernel code.
Could you define BOOST_COMPUTE_DEBUG_KERNEL_COMPILATION at the very beginning of `gpu_tree_learner.h`? Then you will get a detailed OpenCL building log.

**Comment by huanzhang12** [comment]
@Laurae2 I think this is the problem you encountered (Intel OpenCL does not support "cl-strict-aliasing"):
https://github.com/magnumripper/JohnTheRipper/issues/352
If you can confirm this I will just remove this compilation flag.

**Comment by Laurae2** [comment]
@huanzhang12 This might be correct, that flag shows up.

```
(gdb) run
Starting program: C:\github_repos\LightGBM\lightgbm.exe "config=train.conf" "dat
a=binary.train" "valid=binary.test" "objective=binary" "device=gpu"
[New Thread 648.0xd78]
[New Thread 648.0xffc]
[Thread 648.0xffc exited with code 0]
[LightGBM] [Info] Finished loading parameters
[New Thread 648.0xe1c]
[Thread 648.0xe1c exited with code 0]
[LightGBM] [Info] Loading weights...
[New Thread 648.0xdb8]
[Thread 648.0xdb8 exited with code 0]
[New Thread 648.0xe0]
[Thread 648.0xe0 exited with code 0]
[LightGBM] [Info] Loading weights...
[New Thread 648.0x7c8]
[Thread 648.0x7c8 exited with code 0]
[LightGBM] [Info] Finished loading data in 1.093733 seconds
[LightGBM] [Info] Number of positive: 3716, number of negative: 3284
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 6143
[LightGBM] [Info] Number of data: 7000, number of used features: 28
bin size: 255, 256, 256, 255, 256, 255, 256, 256, 3, 255, 256, 256, 3, 255, 256,
 256, 3, 255, 256, 256, 3, 255, 255, 255, 255, 255, 255, 255,
[LightGBM] [Info] Using GPU Device: Intel Xeon E3-12xx v2 (Ivy Bridge), Vendor:
Intel(R) Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
Building GPU kernels with options:  -D POWER_FEATURE_WORKGROUPS=0 -D USE_CONSTAN
T_BUF=0 -D USE_DP_FLOAT=0 -D CONST_HESSIAN=0 -cl-strict-aliasing -cl-mad-enable
-cl-no-signed-zeros -cl-fast-relaxed-math
[New Thread 648.0x660]
[New Thread 648.0xff4]
[New Thread 648.0x970]
[New Thread 648.0xaa0]
[New Thread 648.0x17c]
[New Thread 648.0xf94]
Met Exceptions:
Invalid Program
[Thread 648.0xaa0 exited with code 4294967295]
[Thread 648.0xf94 exited with code 4294967295]
[Thread 648.0x970 exited with code 4294967295]
[Thread 648.0xff4 exited with code 4294967295]
[Thread 648.0x17c exited with code 4294967295]
[Thread 648.0x660 exited with code 4294967295]
[Inferior 1 (process 648) exited with code 037777777777]
```

**Comment by Laurae2** [comment]
I noticed I pasted the wrong log here.. here is the log with the error you described which is the correct one:

```
--- build log ---
Compilation failed
Unrecognized build options: -cl-strict-aliasing

Boost.Compute: terminate called recursively

This application has requested the Runtime to terminate it in an unusual way.
Please contact the application's support team for more information.
Boost.Compute: [Thread 1084.0xf6c exited with code 3]
[Thread 1084.0x5e4 exited with code 3]
[Thread 1084.0x7f8 exited with code 3]
[Thread 1084.0x94c exited with code 3]
[Thread 1084.0x4c0 exited with code 3]
[Thread 1084.0x610 exited with code 3]
[Thread 1084.0x238 exited with code 3]
[Thread 1084.0xd80 exited with code 3]
[Thread 1084.0x664 exited with code 3]
[Thread 1084.0x7e8 exited with code 3]
[Thread 1084.0xfcc exited with code 3]
[Thread 1084.0x79c exited with code 3]
[Thread 1084.0x968 exited with code 3]
[Thread 1084.0xd54 exited with code 3]
[Thread 1084.0x5f8 exited with code 3]
[Thread 1084.0x980 exited with code 3]
[Thread 1084.0x26c exited with code 3]
[Thread 1084.0x970 exited with code 3]
[Thread 1084.0x120 exited with code 3]
[Thread 1084.0xdcc exited with code 3]
[Thread 1084.0xbfc exited with code 3]
[Thread 1084.0xdd4 exited with code 3]
[Thread 1084.0xa9c exited with code 3]
[Thread 1084.0x450 exited with code 3]
[Thread 1084.0xa70 exited with code 3]
[Thread 1084.0x7b0 exited with code 3]
[Thread 1084.0xf10 exited with code 3]
[Thread 1084.0xd7c exited with code 3]
[Thread 1084.0x7d0 exited with code 3]
[Thread 1084.0x954 exited with code 3]
[Thread 1084.0x758 exited with code 3]
[Thread 1084.0x278 exited with code 3]
[Thread 1084.0xf94 exited with code 3]
[Thread 1084.0xd04 exited with code 3]
[Thread 1084.0x2b8 exited with code 3]
[Thread 1084.0x464 exited with code 3]
[Thread 1084.0xec0 exited with code 3]
[Thread 1084.0x3ac exited with code 3]
[Inferior 1 (process 1084) exited with code 03]
```

**Comment by Laurae2** [comment]
@huanzhang12 Do you recommend when using Intel OpenCL to have a special flag for disabling that specific flag, or not? I removed that compilation flag in `gpu_tree_learner.cpp:578` and now it works perfectly.

(I'll close this issue once your PR proposal for the flag will be merged)

**Comment by huanzhang12** [comment]
@Laurae2 In PR #449, I am trying to make the debugging information clearer when such crash happens.
Can you test in on Windows with Intel OpenCL? Ideally it should output something like this:

```
Build Options: -D POWER_FEATURE_WORKGROUPS=0 -D USE_CONSTANT_BUF=0 -D USE_DP_FLOAT=0 -D CONST_HESSIAN=0 -cl-strict-aliasin -cl-mad-enable -cl-no-signed-zeros -cl-fast-relaxed-math
Build Log:
No log available.

[LightGBM] [Fatal] Cannot build GPU program: Invalid Compiler Options
```

Or something like this (depending on the Intel's implementation:

```
Build Options: -D POWER_FEATURE_WORKGROUPS=2 -D USE_CONSTANT_BUF=0 -D USE_DP_FLOAT=0 -D CONST_HESSIAN=0 -cl-strict-aliasing -cl-mad-enable -cl-no-signed-zeros -cl-fast-relaxed-math
Build Log:
Error in processing command line: Don't understand command line argument "-cl-strict-aliasing"!
[LightGBM] [Fatal] Cannot build GPU program: Build Program Failure

```

**Comment by huanzhang12** [comment]
For attribute((max_constant_size(65536))) it will just issue a warning, which is fine. It is just a hint for compilers to generate better code. A compiler can safely ignore it.

**Comment by Laurae2** [comment]
@huanzhang12 Currently I can't test anymore Intel + OpenCL, it seems it crashed my server overnight while doing test runs (I have to reinstall the OS from scratch as it can't boot anymore).

Related to #431, except in my case I cannot boot anymore after the crash.

**Comment by huanzhang12** [comment]
Sorry to know that OpenCL crashed your system... Ideally OpenCL is an user-mode thing; even if the program is buggy (access violations, etc), it will be terminated; the system should not be affected. It seems the Intel OpenCL is a little bit buggy and not as stable as AMD and Nvidia ones.

**Comment by Laurae2** [comment]
@huanzhang12 This is the last run I could do (according to my logs) on that system before it crashed:

* depth = 3
* leaves = 7

It could run without depth limit. But with depth limit it seemed it didn't like and crashed..

`C:\Windows\SYSTEM32\intelopencl64.dll` caused the machine to crash, which caused a critical error on the host kernel crashing the whole machine (watchdog thread lock).

By the way, do you know how to limit the number of cores used using GPU mode? (does `num_threads` apply?)

**Comment by huanzhang12** [comment]
`num_threads` still applies to the CPU part of workload. But there is no way to control how many GPU cores are used.
How large is your dataset? Are you experiencing the same crash on Linux (or GPUs from other vendors on Windows) with this parameter setting? 
Probably some bug in the Intel OpenCL driver is triggered...

**Comment by guolinke** [comment]
I met the same problem.
I run the GPU code in the Intel GPU. And it just freeze. 

log:
```
D:\src\LightGBM\examples\binary_classification>lightgbm.exe config=train.conf device=gpu
[LightGBM] [Info] Finished loading parameters
[LightGBM] [Info] Loading weights...
[LightGBM] [Info] Loading weights...
[LightGBM] [Info] Finished loading data in 0.066501 seconds
[LightGBM] [Info] Number of positive: 3716, number of negative: 3284
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 6143
[LightGBM] [Info] Number of data: 7000, number of used features: 28
[LightGBM] [Info] Using GPU Device: Intel(R) HD Graphics 530, Vendor: Intel(R) Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
1 warning generated.
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 12
[LightGBM] [Info] 28 dense feature groups (0.19 MB) transfered to GPU in 0.003954 secs. 0 sparse feature groups.
[LightGBM] [Info] Finished initializing training
[LightGBM] [Info] Started training...
[LightGBM] [Info] Size of histogram bin entry: 12
[LightGBM] [Info] 28 dense feature groups (0.15 MB) transfered to GPU in 0.002805 secs. 0 sparse feature groups.
```

**Comment by huanzhang12** [comment]
@guolinke Unfortunately I don't have any Windows machines with Intel GPUs to reproduce and investigate this bug. But I think we can try the following way:
- add some debugging printf's in the CPU code and see where it get stalled. I think this line is most possible: https://github.com/Microsoft/LightGBM/blob/master/src/treelearner/gpu_tree_learner.cpp#L197
- If it stalls at that line, there are two possibilities: OpenCL kernel freezes, or the memory copy process (copying histogram back to CPU) freezes. Add kernel_wait_obj_.wait() before line 197, if this line returns, then it is the memory copy problem.
- Otherwise, if  kernel_wait_obj_.wait() does not return, it got stuck at the OpenCL code. Try to add some printf in the OpenCL code and see which part it reaches. The main function is `histogram256`  in https://github.com/Microsoft/LightGBM/blob/master/src/treelearner/ocl/histogram256.cl

Let me know if you are able to find where freeze is, and I will try to figure out what is happening. Thanks!

**Comment by Laurae2** [comment]
@huanzhang12 I am using Bosch dataset, 1M observations, approx 1K sparse features.

@guolinke You can do a small hack the compute module to make it always print the debug because it sometimes crashes but compilation is not error-ing (thus no print):

https://github.com/boostorg/compute/blob/6de7f6448796f67958dde8de4569fb1ae649ee91/include/boost/compute/program.hpp#L257-L268

From there you will be able to see what happened at the end of compilation (did it really compile or not).

**Comment by huanzhang12** [comment]
@Laurae2 It seems in the log of @guolinke the GPU kernels have already successfully built. And it says "started training", so it is probably not a compilation issue. The warnings are about 

For the bosch dataset, do you set `sparse_threshold=1.0`? The full dataset will take about 1 GB GPU memory. How much memory does the run log `xxx dense feature groups (xxx MB) transfered to GPU` say? Probably Intel's GPU driver is buggy when too much memory is used.

**Comment by Laurae2** [comment]
@huanzhang12 I remember only it was taking about 200 MB, using 255 max_bin. The data is 90%+ sparse. I could train with 4k+ leaves without any issue.

**Comment by Laurae2** [comment]
@huanzhang12 I got similar crashes with OpenCL workloads, I think this is due to Intel OpenCL being non standard (works fine with AMD APP SDK, no crash).

We should tell the users to avoid LightGBM + Intel OpenCL on CPU, unless they have a real need for it. Performance is worse than pure CPU in all cases I tested (along with metric performance).

Closing this issue as it seems to be more an Intel OpenCL issue than a LightGBM OpenCL issue. Users should use AMD APP SDK for LightGBM with OpenCL on CPU.
